# Calidad de los Datos.

Despues de descargar y preparar los datos. Procedemos a verificar la calidad de los datos. En esta libreta revisaremos las dimensiones de la calidad de datos.

* Integridad.
* Validez.
* Unicidad.
* Consistencia.
* Actualidad.
* Precisión.

## Origen de Datos

Actualmente poseemos dos fuentes de datos.
1. DGSIAP.
2. Data Mexico.

En esta libreta trabajaremos con el DGSIAP. Pero en el analisis final debemos aplicar todo lo revisado en ambas. 

In [2]:
from scripts.download import agro_project

# Obtenemos los datos establecidos en nuestro analisis.
file_tree = agro_project()


Conexion Establecida
Usted escogio Agricola
Los datos historicos se presentan en el espacio denomiado: Información por nivel
Los datos se encuentran divididos en 3 bloques:

1. Municipal
2. Nacional
3. No seguimiento


Obteniendo los datos en formato de arbol. Ahora procedemos con la descarga.

In [3]:
from scripts.download import load_filetree
# Descargamos los archivos de la estructura de directorios obtenida.
load_filetree(file_tree)

https://nube.agricultura.gob.mx/index.php?view=5D24F995-E734D711-1EA6F8D2-9D1387D7 Diccionario_agricola_2003_a_2025.xlsx /home/dusty/Documents/Repositories/proyecto_agronomia/data/raw/dgsiap /home/dusty/Documents/Repositories/proyecto_agronomia/data/raw/dgsiap/Diccionario_agricola_2003_a_2025.xlsx
El archivo ya fue descargado. Se encuentra en /home/dusty/Documents/Repositories/proyecto_agronomia/data/raw/dgsiap/Diccionario_agricola_2003_a_2025.xlsx
https://nube.agricultura.gob.mx/index.php?view=10AE434F-A2158368-A120BC5A-EDF4AFAA&ANIO=2025 Cierre_agricola_mun_2025.csv /home/dusty/Documents/Repositories/proyecto_agronomia/data/raw/dgsiap /home/dusty/Documents/Repositories/proyecto_agronomia/data/raw/dgsiap/Cierre_agricola_mun_2025.csv
El archivo ya fue descargado. Se encuentra en /home/dusty/Documents/Repositories/proyecto_agronomia/data/raw/dgsiap/Cierre_agricola_mun_2025.csv
https://nube.agricultura.gob.mx/index.php?view=10AE434F-A2158368-A120BC5A-EDF4AFAA&ANIO=2024 Cierre_agricola_mu

### Formateo

Seguiremos con el formateo de los datos. Por el lado de los diccionarios debemos extraer la información relevante en forma de tabla. Mientras que en el caso de los conjuntos de datos debemos resolver el problema de encoding para perder la minima información posible.

In [4]:
from scripts.preparation import format_block, encode_block

blocks= list(file_tree.keys())
encode_project = {b: None for b in blocks}

# Calculando los encodings de los archivos de cada bloque de datos.
encode_project = encode_block(file_tree[blocks[0]])
print(encode_project)


{'Codebook': <scripts.File.File object at 0x7fcca2c61760>, 'Data': {'2025': {'file': <scripts.File.File object at 0x7fcca2c616d0>, 'encoding': 'ISO-8859-1'}, '2024': {'file': <scripts.File.File object at 0x7fcca2c61640>, 'encoding': 'ISO-8859-1'}, '2023': {'file': <scripts.File.File object at 0x7fcca2c615b0>, 'encoding': 'ISO-8859-1'}, '2022': {'file': <scripts.File.File object at 0x7fcca2c61520>, 'encoding': 'ISO-8859-1'}, '2021': {'file': <scripts.File.File object at 0x7fcca2c61490>, 'encoding': 'ISO-8859-1'}, '2020': {'file': <scripts.File.File object at 0x7fcca2c61400>, 'encoding': 'ISO-8859-1'}, '2019': {'file': <scripts.File.File object at 0x7fcca2c61370>, 'encoding': 'ISO-8859-1'}, '2018': {'file': <scripts.File.File object at 0x7fcca2c612e0>, 'encoding': 'ISO-8859-1'}, '2017': {'file': <scripts.File.File object at 0x7fcca2c61250>, 'encoding': 'ISO-8859-1'}, '2016': {'file': <scripts.File.File object at 0x7fcca2c611c0>, 'encoding': 'ISO-8859-1'}, '2015': {'file': <scripts.File.F

Terminando el encoding de los datos procedemos a 
1. Guardar los encodings por si ocupamos volver a descargar los datos no tener que volver a detectar los encodings.
2. Generar los archivos para la calidad de datos.

In [5]:
# Guardar diccionario de encodings en un archivo JSON.
from scripts.config import get_project_root
import pandas as pd

format_project = format_block(encode_project, blocks[0])


ADATA /home/dusty/Documents/Repositories/proyecto_agronomia/data/raw/dgsiap/Diccionario_agricola_2003_a_2025.xlsx


In [6]:
format_project

{'Codebook': {'Diccionario': <scripts.File.File at 0x7fcca2c61f40>,
  'Variables': <scripts.File.File at 0x7fccfc1a16d0>},
 'Data': {'2025': <scripts.File.File at 0x7fccfc1a1d90>,
  '2024': <scripts.File.File at 0x7fccfc1a2720>,
  '2023': <scripts.File.File at 0x7fccfc1a2690>,
  '2022': <scripts.File.File at 0x7fccfc1a24e0>,
  '2021': <scripts.File.File at 0x7fccfc1a1370>,
  '2020': <scripts.File.File at 0x7fccfc1a1f40>,
  '2019': <scripts.File.File at 0x7fccfc1a07a0>,
  '2018': <scripts.File.File at 0x7fccfc1a2840>,
  '2017': <scripts.File.File at 0x7fccfc1a2ba0>,
  '2016': <scripts.File.File at 0x7fccfc1a2e70>,
  '2015': <scripts.File.File at 0x7fccfc1a2f00>,
  '2014': <scripts.File.File at 0x7fccfc1a30b0>,
  '2013': <scripts.File.File at 0x7fccfc1a3650>,
  '2012': <scripts.File.File at 0x7fccfc1a35c0>,
  '2011': <scripts.File.File at 0x7fccfc1a3380>,
  '2010': <scripts.File.File at 0x7fccfc1a3800>,
  '2009': <scripts.File.File at 0x7fccfc1a3890>,
  '2008': <scripts.File.File at 0x7f

In [7]:
import pickle
# Guardar diccionario de encodings en un archivo pickle.
with open(get_project_root() / "docs" / "encode_project.pkl", 'wb') as f:
    pickle.dump(encode_project, f)

with open(get_project_root() / "docs" / "format_project.pkl", 'wb') as f:
    pickle.dump(format_project, f)


# Vemos los archivos formateados del proyecto.


Ahora procedemos con los aspectos de la calidad de datos. 

## Verificación de Datos

En el resto de la libreta iremos viendo cada una de las dimensiones de los datos. En la libreta trabajaremos con un solo archivo pero esto se deberá aplicar para cada uno de las datos a usar para el futuro analisis.

Cargamos los datos previamente formateados. Tomaremos el Codebook asociado y el archivo más reciente.

In [8]:
import pandas as pd


agro_data = format_project
codebook = agro_data['Codebook']

# Cargar el diccionario de datos y sus variables.
catalog = pd.read_excel(codebook['Diccionario'].ubication, index_col=0)
datatypes = pd.read_excel(codebook['Variables'].ubication, index_col=0)

print(catalog.head())
print(datatypes.head())


                 Tipo  Long. Rangos\nválidos  \
Nombre\nde campo                               
Anio                N      4          {CCCC}   
Idestado            N      2       {1,…,32 }   
Nomestado           T     60           {CCC}   
Idddr               N      3       {1,…,193}   
Nomddr              T     60           {CCC}   

                                                        Descripción  \
Nombre\nde campo                                                      
Anio                                          Año de la información   
Idestado          Código que identifica a la entidad federativa ...   
Nomestado          Nombre oficial de la entidad federativa o estado   
Idddr             Código que identifica  al Distrito de Desarrol...   
Nomddr            Nombre que identifica  al Distrito de Desarrol...   

                                                      Observaciones  
Nombre\nde campo                                                     
Anio              Corresp

Notemos que el Diccionario de Datos se compone del catalogo de variables y los tipos de variables definidas. Por ello a lo que usualmente conocemos como diccionario de datos lo estaremos referenciando como catalogo.

En la descarga de datos podemos ver de forma más clara la importancia de la clase File y todo la construcción previa. Solo se hace una vez y facilita los analisis futuros.

In [9]:
# Cargamos la producción elf.link = pathmunicipal de 2025.
data = agro_data['Data']['2025']

# Producción Agricola Municipal 2025
## (Esto se puede deducir siguiendo la construcción del file tree)

### Leemos el archivo
prod_agricola = pd.read_excel(data.ubication)
prod_agricola.head()

,Anio,Idestado,Nomestado,Idddr,Nomddr,Idcader,Nomcader,Idmunicipio,Nommunicipio,Idciclo,...,Nomunidad,Idcultivo,Nomcultivo,Sembrada,Cosechada,Siniestrada,Volumenproduccion,Rendimiento,Preciomediorural,Valorproduccion
0,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,...,Tonelada,5490000,Avena forrajera en verde,1500.0,1500.0,0.0,35250.0,23.50,654.00,23053500.00
1,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,...,Tonelada,5670000,Brócoli,50.0,50.0,0.0,850.0,17.00,6887.78,5854613.00
2,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,2,...,Tonelada,5490000,Avena forrajera en verde,165.0,165.0,0.0,3700.0,22.42,645.00,2386500.00
3,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,2,...,Tonelada,5670000,Brócoli,60.0,60.0,0.0,785.2,13.09,5161.69,4052958.99
4,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,2,...,Tonelada,5740000,Calabacita,30.0,30.0,0.0,490.0,16.33,5285.21,2589752.90


Finalmente, un adelanto del espacio de la validez.

In [10]:
# Esta lista puede modificarse en función de las necesidades
python_types = ["str", "int", "float", "float", "str"]

project_datatypes = {
    t: python_types[k] 
    for k, t in enumerate(datatypes.index)}
    
print(project_datatypes)

{'T': 'str', 'N': 'int', 'D': 'float', '#': 'float', 'A': 'str'}


### Integridad

Pregunta Clave: ¿Los datos están completos y no faltan valores?

Para este paso nos centraremos en dos elementos:
* Coherencia de los datos.
* Datos faltantes o Nulos

#### Coherencia.

Descargamos diccionarios de datos. Esperamos que las columnas coincidan con las variables definidas.

In [11]:
catalog_variables = list(catalog.index)
agro_columns = prod_agricola.columns.to_list()

print("Verifcamos mismo tamaño de columnas", len(catalog_variables) == len(agro_columns))
print("Verificamos mismos elementos", set(catalog_variables) == set(agro_columns))

Verifcamos mismo tamaño de columnas True
Verificamos mismos elementos False


Son el mismo número de elementos. Entonces alguno de las columnas esta escrita distinto.

In [12]:

diff = [(ref, data)
        for ref, data in zip(catalog_variables, agro_columns)
        if ref != data]
print("Columnas distintas")
print(diff)

Columnas distintas
[('Precio', 'Preciomediorural')]


Entnces la columna de Preciomediorural tomo el lugar de la variable Precio en el diccionario. Revisando el diccionario vemos que es la variable precio. 

In [13]:
desc= catalog.loc['Precio']['Descripción']
obsv = catalog.loc['Precio']['Observaciones']
desc, obsv

('Precio medio rural',
 'Precio medio rural, la unidad de medida son pesos por tonelada, con excepción de los cultivos que tienen otra métrica la cual se señala en el nombre del cultivo (gruesa, manojo, planta, entre otros).')

Entonces vemos que el nombre es distinto. Pero la descripción de la columna Precio coincide con el nombre de la columna problemática. Entonces podemos reemplazar las columnas del conjunto de datos con la del diccionario para obtener coherencia entre ambas. 

In [14]:
# Ultima verificación antes de hacer el cambio
for ref, data in zip(catalog_variables, agro_columns):
    print(ref, data)

prod_agricola.columns = catalog_variables

Anio Anio
Idestado Idestado
Nomestado Nomestado
Idddr Idddr
Nomddr Nomddr
Idcader Idcader
Nomcader Nomcader
Idmunicipio Idmunicipio
Nommunicipio Nommunicipio
Idciclo Idciclo
Nomcicloproductivo Nomcicloproductivo
Idmodalidad Idmodalidad
Nommodalidad Nommodalidad
Idunidadmedida Idunidadmedida
Nomunidad Nomunidad
Idcultivo Idcultivo
Nomcultivo Nomcultivo
Sembrada Sembrada
Cosechada Cosechada
Siniestrada Siniestrada
Volumenproduccion Volumenproduccion
Rendimiento Rendimiento
Precio Preciomediorural
Valorproduccion Valorproduccion


Ahora podemos verificar que coinciden, debemos de ver la lista vacia. 

In [15]:
def agro_help(var, guide=catalog):
    assert var in guide.index, "Revisa el nombre de la variable"
    desc= guide.loc[var]['Descripción']
    obsv = guide.loc[var]['Observaciones']
    return {var:{
        "Descripcion": desc,
        "Observaciones": obsv
    }}

agro_columns = prod_agricola.columns.to_list()
diff = [(ref, data)
        for ref, data in zip(catalog_variables, agro_columns)
        if ref != data]
print(diff)

[]


Con los nombres de las columnas adaptados al diccionario (Como un diccionario de datos es inmutable, procuramos que las columnas coincidan con el diccionario, no alrevés) podemos continuar con la segunda parte. 

#### Valores Nulos.
Dentro de la parte de la integridad esta el verificar la existencia de los datos nulos. Puesto que debemos indagar lo suficiente para saber su motivo y como proceder al respecto. Entonces primero debemos ubicarlos.

In [16]:
# Revisamos las columnas
columns_na = prod_agricola.columns[prod_agricola.isna().any()].tolist()
print(columns_na)

['Rendimiento', 'Precio']


Vemos que todos los datos nulos se encuentran en dos columnas. Entonces vamos a centrarnos en esos.

In [17]:
prod_na = prod_agricola[prod_agricola.isna().any(axis=1)]
prod_na.shape

(415, 24)

Vemos que poseemos 415 entradas con valores nulos en una o en las dos columnas previamente revisadas.

In [18]:
prod_na.head(n=8)

,Anio,Idestado,Nomestado,Idddr,Nomddr,Idcader,Nomcader,Idmunicipio,Nommunicipio,Idciclo,...,Nomunidad,Idcultivo,Nomcultivo,Sembrada,Cosechada,Siniestrada,Volumenproduccion,Rendimiento,Precio,Valorproduccion
28,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,3,...,Tonelada,5050000,Agave,8.0,0.0,0.0,0.0,NaN,NaN,0.0
111,2025,1,Aguascalientes,1,Aguascalientes,1,Aguascalientes,11,San Francisco de Los Romo,3,...,Tonelada,7960000,Nuez,79.0,0.0,0.0,0.0,NaN,NaN,0.0
344,2025,2,Baja California,2,Ensenada,1,Tijuana,5,Playas de Rosarito,3,...,Tonelada,9130000,Uva,4.0,0.0,4.0,0.0,NaN,NaN,0.0
367,2025,2,Baja California,2,Ensenada,2,Tecate,3,Tecate,3,...,Planta,8060000,Palma de ornato (planta),2.0,0.0,2.0,0.0,NaN,NaN,0.0
384,2025,2,Baja California,2,Ensenada,3,Ensenada,1,Ensenada,1,...,Tonelada,5480000,Avena forrajera achicalada,4.0,0.0,4.0,0.0,NaN,NaN,0.0
385,2025,2,Baja California,2,Ensenada,3,Ensenada,1,Ensenada,1,...,Tonelada,5510000,Avena grano,1900.0,0.0,1900.0,0.0,NaN,NaN,0.0
386,2025,2,Baja California,2,Ensenada,3,Ensenada,1,Ensenada,1,...,Tonelada,5890000,Cebada forrajera achicalada,15.0,0.0,15.0,0.0,NaN,NaN,0.0
387,2025,2,Baja California,2,Ensenada,3,Ensenada,1,Ensenada,1,...,Tonelada,5920000,Cebada grano,1800.0,0.0,1800.0,0.0,NaN,NaN,0.0


Revisando las primeras columnas vemos algo interesante. Toda la muestra que no posee Rendimiento ni Precio tienen Cosechada y Volumenproduccion en 0.0. El contexto puede darnos una idea del motivo. Pero siempre debemos consultar el diccionario de datos. 

In [19]:
print(agro_help("Sembrada"))
print(agro_help("Cosechada"))
print(agro_help("Siniestrada"))
print(agro_help("Volumenproduccion"))
print(agro_help("Precio"))
print(agro_help("Rendimiento"))
print(agro_help("Valorproduccion"))

{'Sembrada': {'Descripcion': 'Superficie sembrada del cultivo', 'Observaciones': 'La unidad de medida son hectáreas. Es la superficie agrícola en la cual se deposita la semilla de cualquier cultivo, previa preparación del suelo.'}}
{'Cosechada': {'Descripcion': 'Superficie cosechada del cultivo', 'Observaciones': 'La unidad de medida son hectáreas. Es la superficie de la cual se obtuvo producción del cultivo indicado.'}}
{'Siniestrada': {'Descripcion': 'Superficie siniestrada del cultivo', 'Observaciones': 'La unidad de medida son hectáreas. Es la superficie de la cual no se obtuvo producción debido a que se perdió por alguna causa, solo aplica para cultivos anuales, ya que los perennes se pueden recuperar las plantaciones en el año siguiente.'}}
{'Volumenproduccion': {'Descripcion': 'Volumen de producción', 'Observaciones': 'Volumen de producción de la superficie cosechada cuya unidad de medida son las toneladas, con excepción del maguey pulquero y trigo ornamental, con una métrica de

Consultando el diccionario podemos formular una mejor hipotesis para la existencia de los datos. 

* Como el rendimiento se mide con tonelada por hectaria. Podemos deducir que es el cociente de Volumenproduccion y Sembrada. Como ambos son cero obtenemos NaN.
* También, podemos deducir que si Cosechada es 0.0 no se pudo vender y por lo tanto no se registro Precio.

Estas deducciones no tienen porque ser el motivo real. Pero serán hipotesis de las partiremos para el resto del analisis. Pero hay un detalle. ¿Que pasá con el resto de registros?, ¿Cumplen nuestras hipotesis?. Revisemos

In [20]:
# Verificamos si existe registro cuando Volumenproducción no es cero.
check = prod_na[prod_na['Volumenproduccion']!=0.0]
print(check)

Empty DataFrame
Columns: [Anio, Idestado, Nomestado, Idddr, Nomddr, Idcader, Nomcader, Idmunicipio, Nommunicipio, Idciclo, Nomcicloproductivo, Idmodalidad, Nommodalidad, Idunidadmedida, Nomunidad, Idcultivo, Nomcultivo, Sembrada, Cosechada, Siniestrada, Volumenproduccion, Rendimiento, Precio, Valorproduccion]
Index: []

[0 rows x 24 columns]


Vemos que nos devolvio una lista vaica. Entonces todos los valores que posen NaN en Precio y/o Rendimiento poseen Volumenproduccion igual a 0.0. Por lo tanto, nuestras hipotesis se cumplen para todo el conjunto de datos problematico.

**Veredicto** En este caso, bajo nuestras hipotesis, los valores NaN poseen una explicación relevante dentro del conjunto de datos. Se mantendrán pero deben tomarse en cuenta para cualquier pregunta que nos hagamos en el futuro.

### Validez

Pregunta clave:
¿Los datos cumplen con las reglas y restricciones
definidas?

La coherencia entra dentro de este bloque. Sin embargo, se trabajo al comienzo para hacer más comodo el resto del proceso.

La validez se puede resumir en lo siguente que ya tiene el catalogo de datos. 

In [21]:
rules = catalog[['Tipo','Long.','Rangos\nválidos']]
rules

,Tipo,Long.,Rangos\nválidos
Nombre de campo,,,
Anio,N,4,{CCCC}
Idestado,N,2,"{1,…,32 }"
Nomestado,T,60,{CCC}
Idddr,N,3,"{1,…,193}"
Nomddr,T,60,{CCC}
Idcader,N,2,"{1,…,19}"
Nomcader,T,90,{CCC}
Idmunicipio,T,3,"{1,….,570}"
Nommunicipio,T,45,{CCC}


In [22]:
datatypes

,Variable
T,Texto
N,Numérico entero
D,Numérico decimal
#,Carácter numérico.
A,Carácter alfanumérico.


La fuente, el DGSIAP, definió las reglas que cada variable debe seguir.

* Tipo de Dato.
* Longitud.
* Rangos Validos.

Entonces debemos verificar cada uno de puntos. Sin embargo.

In [23]:
# Tipo de Dato.
prod_dtypes = prod_agricola.dtypes
print(prod_dtypes)

Anio                    int64
Idestado                int64
Nomestado                 str
Idddr                   int64
Nomddr                    str
Idcader                 int64
Nomcader                  str
Idmunicipio             int64
Nommunicipio              str
Idciclo                 int64
Nomcicloproductivo        str
Idmodalidad             int64
Nommodalidad              str
Idunidadmedida          int64
Nomunidad                 str
Idcultivo               int64
Nomcultivo                str
Sembrada              float64
Cosechada             float64
Siniestrada           float64
Volumenproduccion     float64
Rendimiento           float64
Precio                float64
Valorproduccion       float64
dtype: object


En general no nos interesa la precisión del entero y/o el flotante (el número despues de int/float). Entonces solo revisaremos que forme parte de los tipos de dato elemental.

In [24]:
for col in agro_columns:
    col_type = rules.loc[col, 'Tipo']
    col_dtype =str(project_datatypes[col_type])
    ac_dtype = str(prod_dtypes[col])
    try:
        ac_dtype.index(col_dtype)
        project_datatypes[col_type] = prod_dtypes[col]
    except:
        print(col, ac_dtype, col_dtype)

Idmunicipio int64 str


Como podemos ver la columna Idmunicipio viene definida como un entero pero el diccionario de datos lo marco como texto (T). Pero se nos presenta un dilema. 

In [25]:
rules.loc[[r for r in rules.index if r[:2] == "Id"]]

,Tipo,Long.,Rangos\nválidos
Nombre de campo,,,
Idestado,N,2,"{1,…,32 }"
Idddr,N,3,"{1,…,193}"
Idcader,N,2,"{1,…,19}"
Idmunicipio,T,3,"{1,….,570}"
Idciclo,N,1,"{1,….,3}"
Idmodalidad,N,1,"{1,2}"
Idunidadmedida,N,1,"{1,….,6}"
Idcultivo,N,3,"{1,….,410}"


Vemos que Idmunicipio es el único identificador que es texto, el resto son enteros. 

**Dilema** Hace un momento comentamos que el diccionario es inmutable. Sin embargo, considerando que Idestado es un entero, podemos deducir que no hay un argumento fuerte para que sea un elemento de texto y habra sido un error de dedo.

**Veredicto** No vamos a modificar el diccionario de datos. Sin embargo, tomando en cuenta que para el analisis posterior convertiremos los datos Id enteros a categoricos no afecta al analisis. Pero esto debe ser mencionado en el resumen de la calidad de los datos.

Con los tipos de datos corroborado ahora procedemos a la parte de los rangos. Para esto lo haremos de forma natural y luego probaremos con la libreria pandatic.

In [26]:
# Recordamos nuestros tipos de datos del proyecto.
project_datatypes

{'T': <StringDtype(storage='python', na_value=nan)>,
 'N': dtype('int64'),
 'D': dtype('float64'),
 '#': 'float',
 'A': 'str'}

#### Variables Categoricas
En papel de momento. Pero a eso iremos.

In [27]:
import re


def long_rule(val, long):
    assert str(val), "Revisa la entrada"
    return len(str(val)) <= long

def range_rule(val, rage: tuple):
    a, b = rage
    check_set = set(range(a, b+1))
    return int(val) in check_set


# Reglas de Variables categoticas
## Caso particular 
rl_mun = rules.loc['Idmunicipio']
long = rl_mun['Long.']
range_mun = rl_mun["Rangos\nválidos"]
datarules = {"Idmunicipio": lambda x, l=long: long_rule(x, l)}


cat_vars = rules[rules['Tipo']=="N"]
cat_vars.drop(axis=0, index="Anio", inplace=True)
for var in cat_vars.index:
    k = re.search(r"{(.*)}",cat_vars.loc[var, "Rangos\nválidos"])
    ranges = k.group(1)
    g_min = ranges.split(",")[0]
    g_max = ranges.split(",")[-1]
    datarules[var] = lambda x, ival=(int(g_min), int(g_max)): range_rule(x, ival)



#### Variables de Texto y Variables Numéricas

In [28]:
text_vars = rules[rules['Tipo']=="T"]
text_vars.drop(axis=0, index="Idmunicipio", inplace=True)
text_vars

,Tipo,Long.,Rangos\nválidos
Nombre de campo,,,
Nomestado,T,60,{CCC}
Nomddr,T,60,{CCC}
Nomcader,T,90,{CCC}
Nommunicipio,T,45,{CCC}
Nomcicloproductivo,T,40,{CCC}
Nommodalidad,T,20,{CCC}
Nomunidad,T,40,{CCC}
Nomcultivo,T,40,{CCC}


In [29]:
# Reglas para variables de texto
for var in text_vars.index:
    # print(var,text_vars.loc[var, "Long."])
    datarules[var] = lambda x, l = text_vars.loc[var, "Long."]: long_rule(x, l)


# Reglas para variables numéricas.
def interval_rule(val, interval: tuple, config=(True, True)):
    val = float(val)
    try:
        a, b = interval
        if config[0] :
            if config[1]:
                return a <= val <= b
            else:
                return a <= val < b
        else:
            if config[1]:
                return a < val <= b
            else:
                return a < val < b
    except:
        print(val)
        print(f"SI {val} es un valor nulo. Debe revisar el contexto")
        return 1


# Caso particular. El año.
rl_year = rules.loc['Anio']
nd = rl_year['Long.']
my = 0
mmy = 10 ** nd
datarules['Anio'] = lambda x, ival=(my, mmy): interval_rule(x, ival, config=(True, False))


num_vars = rules[rules['Tipo']=="D"]

for var in num_vars.index:
    k = re.search(r"{(.*)}",num_vars.loc[var, "Rangos\nválidos"])
    ranges = k.group(1)
    dec = len(ranges.split(".")[-1])
    nd = num_vars.loc[var, "Long."]
    m = 0
    mm = 10 ** nd - 1
    datarules[var] = lambda x, ival=(m, mm): interval_rule(x, ival)
    

print(datarules)

{'Idmunicipio': <function <lambda> at 0x7fcc957216f0>, 'Idestado': <function <lambda> at 0x7fcc95720f60>, 'Idddr': <function <lambda> at 0x7fcc95721220>, 'Idcader': <function <lambda> at 0x7fcc957217a0>, 'Idciclo': <function <lambda> at 0x7fcc95721640>, 'Idmodalidad': <function <lambda> at 0x7fcc957219b0>, 'Idunidadmedida': <function <lambda> at 0x7fcc95721a60>, 'Idcultivo': <function <lambda> at 0x7fcc95721b10>, 'Nomestado': <function <lambda> at 0x7fcc95721c70>, 'Nomddr': <function <lambda> at 0x7fcc95721e80>, 'Nomcader': <function <lambda> at 0x7fcc95721bc0>, 'Nommunicipio': <function <lambda> at 0x7fcc95721f30>, 'Nomcicloproductivo': <function <lambda> at 0x7fcc95721d20>, 'Nommodalidad': <function <lambda> at 0x7fcc95721fe0>, 'Nomunidad': <function <lambda> at 0x7fcc95722090>, 'Nomcultivo': <function <lambda> at 0x7fcc95722140>, 'Anio': <function <lambda> at 0x7fcc957222a0>, 'Sembrada': <function <lambda> at 0x7fcc95722560>, 'Cosechada': <function <lambda> at 0x7fcc957224b0>, 'Sini

Con eso ya tenemos un conjunto de reglas para cada columna de nuestro conjunto de datos. Ahora revisamos si las columnas satifacen las reglas previamente establecidas.

In [30]:
# print(prod_agricola.head(n=5))

bad_columns = []

for col in prod_agricola.columns:
    check_rule = datarules[col]
    validate = prod_agricola[col].apply(check_rule)
    if not validate.all():
        bad_columns.append(col)
        print(f"Error en la columna {col}")

Error en la columna Nommunicipio
Error en la columna Idunidadmedida
Error en la columna Idcultivo
Error en la columna Rendimiento
Error en la columna Precio


Vemos que hay 5 columnas que por lo menos poseen un valor que no satisface las reglas establecidas por el diccionario. Revisando las columnas podemos ver los elementos que no cumplen las reglas.

In [31]:
col = bad_columns[0]

# El nombre del municipio no debe exceder los 45 caracteres.
print(rules.loc[col])

# Filtrar las filas que no cumplen con la regla
fail_col = prod_agricola[col]
fail_col = fail_col[~fail_col.apply(datarules[col])]
fail_df = pd.DataFrame(fail_col)
fail_df['Long'] = fail_col.apply(lambda x: len(str(x)))
fail_rate = len(fail_df) / len(prod_agricola) * 100
print("Valores fuera de regla", len(fail_df))
fail_rate

Tipo                   T
Long.                 45
Rangos\nválidos    {CCC}
Name: Nommunicipio, dtype: object
Valores fuera de regla 63


0.175477689265222

Viendo entonces que la primer columna posee 63 valores fuera de la regla definida por el diccionario. Que en general es un 0.17%. En este caso no es algo grave pero es algo que se debe comentar en el resumen de calidad de datos.

Ahora haremos el mismo proceso pero con el resto de columnas con fallos.

In [32]:
rules

,Tipo,Long.,Rangos\nválidos
Nombre de campo,,,
Anio,N,4,{CCCC}
Idestado,N,2,"{1,…,32 }"
Nomestado,T,60,{CCC}
Idddr,N,3,"{1,…,193}"
Nomddr,T,60,{CCC}
Idcader,N,2,"{1,…,19}"
Nomcader,T,90,{CCC}
Idmunicipio,T,3,"{1,….,570}"
Nommunicipio,T,45,{CCC}


In [33]:
def check_bad_columns(cols):
    bad_data = {}
    for c in cols:
        fail_col = prod_agricola[c]
        fail_col = fail_col[~fail_col.apply(datarules[c])]
        fail_rate = len(fail_col) / len(prod_agricola) * 100
        bad_data[c] = fail_col
        # print(c, rules.loc[c])
        print(f"Columna {c} posee {len(fail_col)} valores fuera de regla")
        print(f"Obteniendo una tasa de fallo de ({fail_rate:.2f}%)")
        print(f"\n")
    return bad_data

baddata = check_bad_columns(bad_columns)

Columna Nommunicipio posee 63 valores fuera de regla
Obteniendo una tasa de fallo de (0.18%)


Columna Idunidadmedida posee 35902 valores fuera de regla
Obteniendo una tasa de fallo de (100.00%)


Columna Idcultivo posee 35902 valores fuera de regla
Obteniendo una tasa de fallo de (100.00%)


Columna Rendimiento posee 415 valores fuera de regla
Obteniendo una tasa de fallo de (1.16%)


Columna Precio posee 415 valores fuera de regla
Obteniendo una tasa de fallo de (1.16%)




Vemos que tenemos la columna de nombres de municipios. Donde se quedaron cortos con el nombre de los municipios. Revisando el resto vemos algo interesante

In [34]:
# Recordamos la regla de la columna Idcultivo
print(rules.loc['Idcultivo'])

baddata['Idcultivo']

Tipo                        N
Long.                       3
Rangos\nválidos    {1,….,410}
Name: Idcultivo, dtype: object


0         5490000
1         5670000
2         5490000
3         5670000
4         5740000
           ...   
35897     7960000
35898     9130000
35899    15050000
35900     5050000
35901    15050000
Name: Idcultivo, Length: 35902, dtype: int64

In [35]:
agro_help("Idcultivo")

{'Idcultivo': {'Descripcion': 'Código que identifica al cultivo agrícola',
  'Observaciones': 'Clave asignado en el catálogo de cultivos de la DGSIAP.'}}

Viendo que en este caso la variable esta incompleta. En el sentido que el diccionario dice que tiene a lo más 3 digitos. Pero no explica porque en la base de datos escribe 4 ceros al comienzo.

In [36]:
# Recordamos la regla de la columna Idcultivo
print(rules.loc['Idunidadmedida'])

baddata['Idunidadmedida']

Tipo                      N
Long.                     1
Rangos\nválidos    {1,….,6}
Name: Idunidadmedida, dtype: object


0        200201
1        200201
2        200201
3        200201
4        200201
          ...  
35897    200201
35898    200201
35899    200201
35900    200201
35901    200201
Name: Idunidadmedida, Length: 35902, dtype: int64

In [37]:
agro_help("Idunidadmedida")

{'Idunidadmedida': {'Descripcion': 'Código que identificala unidad de medida',
  'Observaciones': 'Código asignado en el catálogo de unidad de medida.'}}

Vemos que pasa algo similar a Idcultivo. Información incompleta.

In [38]:
print(rules.loc['Precio'])

print(baddata['Precio'].isna())
# Quitamos los valores nulos.
print(baddata['Precio'][~baddata['Precio'].isna()])

Tipo                           D
Long.                          7
Rangos\nválidos    {#######.###}
Name: Precio, dtype: object
28       True
111      True
344      True
367      True
384      True
         ... 
35798    True
35805    True
35856    True
35857    True
35890    True
Name: Precio, Length: 415, dtype: bool
Series([], Name: Precio, dtype: float64)


Como vemos. Los unicos datos que no cumplen las reglas son los valores nulos. Que ya hablamos en la parte de Integridad que son importantes para el analisis. Por lo tanto, es un fallo a la regla. Pero es aceptable. Bajo la misma premisa ocurrira lo mismo con el rendimiento. 

In [39]:
print(rules.loc['Rendimiento'])

print(baddata['Rendimiento'].isna())
# Quitamos los valores nulos.
print(baddata['Rendimiento'][~baddata['Rendimiento'].isna()])

Tipo                            D
Long.                          10
Rangos\nválidos    {########.###}
Name: Rendimiento, dtype: object
28       True
111      True
344      True
367      True
384      True
         ... 
35798    True
35805    True
35856    True
35857    True
35890    True
Name: Rendimiento, Length: 415, dtype: bool
Series([], Name: Rendimiento, dtype: float64)


#### Variante usando pandera.

Ahora que ya vimos una idea general de como se hace. Para el proyecto lo haremos usando Pandera y su complemento para pandas.

In [48]:
import pandera.pandas as pa

agro_schema = pa.DataFrameSchema({
    col: pa.Column(prod_agricola[col].dtype, pa.Check(datarules[col]), nullable=True)
    },
    index = pa.Index(int),
    name="dgsiap_mun"
)

try:
    agro_schema.validate(prod_agricola, lazy=True)
except pa.errors.SchemaErrors as exc:
    print(exc)

{
    "DATA": {
        "DATAFRAME_CHECK": [
            {
                "schema": "dgsiap_mun",
                "column": "Nommunicipio",
                "check": "<lambda>",
                "error": "Column 'Nommunicipio' failed series or dataframe validator 0: <Check <lambda>>"
            }
        ]
    }
}


Viendo como La columna municipio posee un problema. Algo que ya habiamos revisado en la revisión pasada.

### Unicidad 

Pregunta clave
¿Hay datos duplicados?

Ya vimos que los datos tienen problemas de Integridad (Existen valores nuloes) y problemas de validez (Variables no siguen las reglas definidas por el diccionario de datos). Ahora sigue revisar si el conjunto de datos posee valores duplicados.

In [53]:
len(prod_agricola.drop_duplicates())-len(prod_agricola)

0

Viendo que no existe valores duplicados.

### Consistencia

Pregunta clave
¿Los datos son coherentes entre diferentes fuentes
y representaciones?

Luego sigue la consistencia. Muchas veces una variable puede obtenerse de distintas fuentes. Entonces debemos asegurarnos que la variable mantenga su valor de forma indistinta. SIn embargo en este caso, como la información proviene de su emisora solo la revisaremos cuando aparezca en distintas fuentes de datos.

### Actualidad
¿Los datos están actualizados y reflejan la realidad
actual?

La necesidad de actualidad es varaible según el problema. Esto por la necesidad de los datos historicos. Sin embargo, siempre debe de realizarse en analisis de la demanda de los datos. 

En nuestro proyecto requerimos datos historicos para realizar una inferencia. Sin embargo, en nuestro caso es importante poseer los datos de los años más recientes para mejor calidad de la inferencia. 

### Precisión

¿Los datos son correctos y libres de errores?

Finalmente llegamos a la precisión. Apartado donde ocupamos ver si los datos tiene errores con respecto a otras fuentes de datos, por lo general más fiables. En este caso la fuente de los datos es la emisora de los datos.

## Conclusiones.

Despues de toda la revisión de la calidad de los datos sacamos las siguientes conclusiones.

* Los datos posee problemas de Integridad.
    * Una columna con un nombre que no corresponde con el diccionario.
    * Una columna con valores nulos.
* Problemas de Validez.
    * Hay una columna con el tipo de datos incorrecto.
    * Varias columnas no siguen las reglas establecidas por el diccionario. 
* Problemas de Unicidad.
    * No posee datos duplicados.
* Al ser la fuente de datos oficial.
    * Los datos no tiene problemas de consistencia.
    * Los datos no tiene problemas de actualidad.
    * Los datos no tiene problemas de actualidad.